### PR TITLE
Rework docker image deletion

### PR DESCRIPTION
**Pull request by nikola-rados**
# Overview
This PR reworks how the `jenkins-library` handles image names in the building, publishing and deleting of docker images.  Three steps were updated (`buildDockerImage`, `publishDockerImage`, `removeDockerImage`), one was created (`buildImageName`) and one was removed (`getScanName`).  These `steps` have been testing live using a test `climate-explorer-backend` [pipeline](https://pcic.uvic.ca/jenkins/job/PCIC/job/climate-explorer-backend/job/docker-testing/).  The class functions have been tested using `gradle`.

Once this PR is merged I was thinking of tagging it, would `0.1.0` be appropriate?

Resolves issue #10. 

# Background
This section will outline some of the discoveries I made in solving this issue.  **TL;DR at the bottom**

## Getting Image Names
Previously, the construction of a docker image name was being handled inside of `buildDockerImage`.  The `imageName` would be returned in a list alongside the image object.  There may be cases where we want the image name for another `step`, but we have no run build yet.  Because of this, it made sense to pull this functionality out of build and create a separate `step`.  `buildImageName` constructs the image name (default is `pcic/[base]:[branch-name]`) and is more flexible since each component (`registry`, `base` and `tag`) is a parameter.  This `step` makes `getScanName` obsolete.

## Incorrect Initial Build Tag
`buildDockerImage` was **not** applying a tag during the build step.  This intentional as I wanted the tagging to be handled by the publishing step.  As it turns out, if no tag is given during the build, docker will automatically tag the image with `latest`.  Going forward, `imageName` should include a tag (preferably the one produced by `buildImageName`).

## `<none>` Tagged Images
For some time now we have been finding images tagged with `<none>`.  It looks like tagging an image multiple times may be the cause of this (which we do in our publishing step).  Our old deletion strategy did not cover this case, hence the build up.

## Bad Image Deletion
The old deletion strategy would only delete images tagged with `latest`, this was **not** intentional.  The updated strategy deletes based on `name:tag`.  This allows us to delete the base image and any of the tags it may have (this was a source of build up in `dev01`).  Furthermore, since we are tagging an image multiple times there is a chance a `<none>` tag appears.  These tags cannot be deleted using the `name:tag` strategy and thus we must delete using the image `id`.  `removeDockerImage` handles this in a `try/catch` block were it attempts to delete by `id`.  If there is no `<none>` image an exception will be thrown that we display but ignore.

**TL;DR** - Some unexpected behaviour from the build `step` was causing issues in the publish and remove steps.

Edit: Added sentence about tests in Overview.

### CODE REVIEW

File: src/org/pcic/DockerUtils.groovy [hunk 1]
```diff
@@ -1,37 +1,54 @@
...
 
     DockerUtils(steps) {
         this.steps = steps
+        this.utils = new Utils(steps)
     }
 
-    void removeImage(String imageName) {
-        steps.sh(script: "docker rmi ${imageName}")
+    void removeImage(String image) {
+        steps.sh(script: "docker rmi ${image}")
```
Comment (rod-glover) [inline src/org/pcic/DockerUtils.groovy:18]: I kinda like `imageName`; an image and its name are distinct things. That said, if this is more in line with your general pattern of identifiers, then stick with `image`.

File: src/org/pcic/DockerUtils.groovy [hunk 2]
```diff
@@ -1,37 +1,54 @@
...
     }
 
-    String buildImageName(String suffix) {
-        if (suffix.contains('/')) {
-            throw new Exception("Image name suffix: ${suffix} cannot contain `/` character")
+    String formatDockerString(String str, String replacement = '-') {
+        for (String illegalChar : Global.constants.illegalDockerChars) {
+            if (str.contains(illegalChar)) {
```
Comment (rod-glover) [inline src/org/pcic/DockerUtils.groovy]: Maybe we don't have to do this test? Just run the `replaceAll` and if there is nothing to replace it's a noop?

File: src/org/pcic/DockerUtils.groovy [hunk 3]
```diff
@@ -1,37 +1,54 @@
...
+                str = str.replaceAll(illegalChar, replacement)
+            }
         }
-        return steps.env.BASE_REGISTRY + suffix
+
+        return str
+    }
+
+    String buildImageName(String registry, String base, String tag) {
+        return formatDockerString(registry) + '/' + formatDockerString(base) + ":" + formatDockerString(tag)
```
Comment (rod-glover) [inline src/org/pcic/DockerUtils.groovy:30]: Nice

File: src/org/pcic/GitUtils.groovy [hunk 4]
```diff
@@ -8,14 +8,14 @@ class GitUtils implements Serializable {
         this.steps = steps
     }
 
-    ArrayList isCommitTagged() {
-        String gitTags = steps.sh(script: 'git tag --contains', returnStdout: true)
+    ArrayList getCommitTags() {
+        String tags = steps.sh(script: 'git tag --contains', returnStdout: true)
 
-        if (gitTags.isEmpty()) {
+        if (tags.isEmpty()) {
```
Comment (rod-glover) [inline src/org/pcic/GitUtils.groovy:14]: Sorry, I have to ask: Doesn't `tags.trim().split('\n')` work properly when `tags.isEmpty()`?

**Review by rod-glover — APPROVED** [review]
Great work. The code seems clearer and cleaner to me. Tests are very nice, evidence of a good design. Only a couple of minor coding questions.

File: src/org/pcic/DockerUtils.groovy (hunk 1 above)
Comment (nikola-rados) [reply src/org/pcic/DockerUtils.groovy:18]: The reason I changed it to `image` from `imageName` is because it doesn't necessarily have to be the name of the image (it could be the `image_id` as well).

File: src/org/pcic/GitUtils.groovy (hunk 4 above)
Comment (nikola-rados) [reply src/org/pcic/GitUtils.groovy:14]: Unfortunately in `groovy` an error is thrown if you run `.trim()` on an empty string.

**Comment by rod-glover** [comment]
Merge at will

File: src/org/pcic/DockerUtils.groovy (hunk 2 above)
Comment (nikola-rados) [reply src/org/pcic/DockerUtils.groovy]: Good idea!

**Comment by nikola-rados** [comment]
@rod-glover did you have any comment about the release version?

**Comment by rod-glover** [comment]
I'm not sure what you mean by that.
